build/nested-cluster/start.sh

```diff
@@ -0,0 +1,121 @@
+#!/bin/bash
+
+# Copyright 2017 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+wait_for_docker ()
+{
+  # Wait for docker.
+  until docker version; do sleep 1 ;done
```

Comment:
This doesn't seem like a very strong contract to me. If this is to be used for testing, please consider finding a better solution.

Reply:
This is more or less fine since tests timeout through other means. It's probably worth having this eventually fail but FWIW we use DinD for testing right now and do essentially this after discussing possible solutions with the GKE node team. https://github.com/kubernetes/test-infra/blob/9bc72b9fb345974ded6a24cd3fcfae4038cc7d2b/images/bootstrap/runner#L36-L51

If there's a better solution I'd like to know it.

Reply:
Declaring the dependency via systemd would be a start, e.g. `After=docker.service`. Once that's establish, the chances are we aren't gonna have to wait.

Reply:
We plan to clean up the systemd dependencies in the future, but the first iteration doesn't need to be perfect.

Reply:
Makes sense.

Reply:
> Declaring the dependency via systemd would be a start, e.g. After=docker.service. Once that's establish, the chances are we aren't gonna have to wait. This would allow to fail early, when systemd cannot even start docker.

In the debian image with DinD in test infra at least (not-systemd) the init script success and does indeed start docker but for whatever inexplicable reason dockerd isn't ready to talk over the socket yet. 

Checking for `docker ps` to exit cleanly with some backoff and an eventual timeout works well FWIW. We're doing this on every k8s PR now with good success.